### PR TITLE
raftstore: delete all cfs' data before applying snapshot.

### PR DESCRIPTION
**Pull request by disksing**
@BusyJay @zhangjinpeng1987 @ngaut @siddontang

### CODE REVIEW

File: tests/raftstore/test_snap.rs [hunk 1]
```diff
@@ -203,3 +203,52 @@ fn test_server_concurrent_snap() {
...
+    cluster.run();
+    let cf = "lock";
+    cluster.must_put_cf(cf, b"k1", b"v1");
+    cluster.must_put_cf(cf, b"k2", b"v2");
+    let engine1 = cluster.get_engine(1);
+    must_get_cf_equal(&engine1, cf, b"k1", b"v1");
+    must_get_cf_equal(&engine1, cf, b"k2", b"v2");
+
+    // Isolate node 1.
+    cluster.stop_node(1);
```
Comment (siddontang) [inline tests/raftstore/test_snap.rs]: you can use filter to isolate node.

File: tests/raftstore/test_snap.rs [hunk 2]
```diff
@@ -203,3 +203,52 @@ fn test_server_concurrent_snap() {
...
+    }
+
+    cluster.must_delete_cf(cf, b"k2");
+
+    // Add node 1 back.
+    cluster.run_node(1);
+
+    cluster.must_put_cf(cf, b"k3", b"v3");
+    must_get_cf_equal(&engine1, cf, b"k3", b"v3");
+    // Now snapshot must be applied on node 2.
```
Comment (siddontang) [inline tests/raftstore/test_snap.rs]: node 2? mismatch comment?

**Comment by siddontang** [comment]
LGTM

File: src/raftstore/store/peer_storage.rs [hunk 3]
```diff
@@ -376,11 +377,15 @@ impl PeerStorage {
             // we can only delete the old data when the peer is initialized.
             let timer = Instant::now();
             // Delete everything in the region for this peer.
-            try!(self.scan_region(self.engine.as_ref(),
-                                  &mut |key, _| {
-                                      try!(ctx.wb.delete(key));
-                                      Ok(true)
-                                  }));
+            for cf in self.engine.as_ref().cf_names() {
```
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Why as_ref?

File: src/raftstore/store/peer_storage.rs [hunk 4]
```diff
@@ -480,6 +485,18 @@ impl PeerStorage {
         Ok(())
     }
 
+    pub fn scan_region_cf<T, F>(&self, db: &T, cf: &str, f: &mut F) -> Result<()>
+        where T: Iterable,
+              F: FnMut(&[u8], &[u8]) -> Result<bool>
+    {
+        let ranges = self.region_key_ranges();
```
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Not all range need to be scan.

File: src/raftstore/store/peer_storage.rs [hunk 5]
```diff
@@ -431,6 +425,37 @@ impl PeerStorage {
...
+                                  &mut |key, _| {
+                                      try!(wb.delete(key));
+                                      Ok(true)
+                                  }));
+        }
+
+        let (start_key, end_key) = (enc_start_key(self.get_region()),
+                                    enc_end_key(self.get_region()));
+        for cf in self.engine.cf_names() {
+            if cf == "default" {
```
Comment (BusyJay) [inline src/raftstore/store/peer_storage.rs]: Can we use `DEFAULT_CFNAME` here?

**Comment by BusyJay** [comment]
rest LGTM

**Pull request merged**